lib/alphavantage/client.rb

```diff
@@ -1,27 +1,62 @@
+require 'securerandom'
+require 'json'
+require 'csv'
+
 module Alphavantage
   class Client
 
     class << self
-      def get(params:)
-        default_params = { apikey: Alphavantage.configuration.api_key }
-        response = Faraday.get('https://www.alphavantage.co/query') do |req|
-          req.params = default_params.merge(params)
-          req.headers['Content-Type'] = 'application/json'
-        end
-        Hashie::Mash.new(convert_hash_keys(JSON.parse(response.body)))
+      def get(params:, datatype: :json)
+        new(params).public_send(datatype)
+      end
+    end
+
+    def initialize params
+      @params = params
+    end
+    attr_reader :params
+
+    def json
+      Hashie::Mash.new(convert_hash_keys(JSON.parse(response.body))).tap do |response|
+        raise Error, response.error_message if response.error_message
       end
+    end
 
-      def convert_hash_keys(value)
-        case value
-          when Array
-            value.map { |v| convert_hash_keys(v) }
-          when Hash
-            Hash[value.map { |k, v| [ NormalizeKey.new(key: k).call, convert_hash_keys(v) ] }]
-          else
-            value
-         end
+    def csv
+      CSV.parse response.body
+    rescue CSV::MalformedCSVError
+      # if we can not parse it, we probably have JSON from API with an error
+      json
+      raise
```

Comment:
@kvokka don't think you need this line since calling `json` would already be raising the Error

Reply:
In the best world- yes, this raise will be unreachable. But, if, for some unknown reason we get just an odd response (which might happen in case of high load of the service or some breaking update) this will be raised for sure.

I prefer to treat the quotes service as a black box, so it may do anything.
wdyt?